proxy/src/proxy.rs

```diff
@@ -124,8 +102,19 @@ pub async fn task_main(
                     }),
                 );
             }
-            _ = PROXY_CANCELLATION_TOKEN.cancelled() => {
-                bail!("Exiting");
+            maybe_exited = connections.join_next() => {
+                match maybe_exited
+                {
+                    Some(Ok(())) => {},
+                    Some(Err(e)) => { bail!(e.to_string()) },
+                    None => {
+                        if cancellation_token.is_cancelled()
+                        {
+                            drop(listener);
+                            bail!("All connections closed, exiting");
```

Comment:
I'd call this very much a normal exit assuming this is how the updates will done, no error needed, just `return Ok(())` or break out of the loop.

Reply:
I think we actually need an error here because we want Tokio to shut everything down when this loop exits

Reply:
That is correct, didn't check the main. Resolving..